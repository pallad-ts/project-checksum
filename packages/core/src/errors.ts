import {Domain, generators} from "alpha-errors";
import {ChecksumError} from "./ChecksumError";

export const ERRORS = Domain.create({
	errorClass: ChecksumError,
	codeGenerator: generators.formatCode('E_PROJECT_CHECKSUM_%d')
})
	.createErrors(create => {
		return {
			NO_CONFIG_FOUND: create('No @pallad/project-checksum configuration found in directory: %s'),
			NO_CONFIG_FILE_FOUND: create('No config file "%s" for @pallad/project-checksum found in directory: %s'),
			NO_LOADER_FOUND: create('No loader "%s" found'),
			NO_LOADER_EXPORTED: create('No loader exported in "%s" package'),
			NO_CONTENT_TO_HASH_COMPUTE_FOUND: create("Project does not have any content to compute checksum from"),
			CIRCULAR_DEPENDENCY: create('Circular dependency detected. Projects path: %s')
		}
	});
