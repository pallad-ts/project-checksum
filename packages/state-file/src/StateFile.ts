import {byJsonSchema} from "alpha-validator-bridge-jsonschema";
import * as fs from 'fs/promises';
import * as YAML from 'yaml';
import {ValidatorError} from "alpha-validator";
import {State} from "./types";
import {just, Maybe, none} from "@sweet-monads/maybe";

const schema = require('../state-schema.json');

const validator = byJsonSchema<Record<string, string>>(schema);

export namespace StateFile {
	export async function load(file: string): Promise<Maybe<State>> {
		const isFileExist = await fs.access(file).then(() => true, () => false);
		if (!isFileExist) {
			return none();
		}
		const content = await fs.readFile(file, 'utf-8');

		const data = YAML.parse(content);
		const state = validate(data);

		if (state.isLeft()) {
			throw new ValidatorError(state.value, 'Invalid state file structure');
		}

		return just(state.value);
	}

	export function validate(data: any) {
		return validator(data)
			.map((result: Record<string, string>) => {
				const map = new Map<string, string>();
				for (const [projectName, checksum] of Object.entries(result)) {
					map.set(projectName, checksum);
				}
				return map;
			});
	}

	export async function save(file: string, state: State) {
		const data = Object.fromEntries(state.entries());
		const content = YAML.stringify(data);
		await fs.writeFile(file, content);
	}
}
