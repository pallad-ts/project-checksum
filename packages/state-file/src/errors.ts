import {Domain, generators} from "alpha-errors";
import {ChecksumError} from "@pallad/project-checksum-core";

export const ERRORS = Domain.create({
	errorClass: ChecksumError,
	codeGenerator: generators.formatCode('E_PROJECT_CHECKSUM_STATE_%d')
})
	.createErrors(create => {
		return {
			NO_PROJECT_FOUND: create('No checksum project "%s" found'),
			STATE_FILE_NOT_FOUND: create('No state file "%s" found'),
			NO_PROJECT_CHECKSUM_IN_STATE: create('No checksum in state for project "%s"'),
			NO_CONFIG_FOUND: create('No configuration file found in directory "%s"'),
			CONFIG_FILE_NOT_FOUND: create('Configuration file "%s" not found')
		}
	});
