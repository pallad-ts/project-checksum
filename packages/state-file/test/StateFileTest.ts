import {StateFile} from "@src/StateFile";
import * as path from "path";
import {ERRORS} from "@src/errors";
import * as fs from 'fs';
import {left, right} from "@sweet-monads/either";
import {just} from "@sweet-monads/maybe";

describe('StateFile', () => {
	describe('validation', () => {
		it('success', () => {
			const result = StateFile.validate({
				this: 'isfine'
			});

			expect(result.unwrap())
				.toEqual(
					new Map([['this', 'isfine']])
				);
		});

		it('invalid data', () => {
			const result = StateFile.validate({
				this: {is: 'notok'}
			});

			expect(result.value)
				.toMatchSnapshot();
		});
	});

	describe('loading', () => {
		it('success', async () => {
			const data = await StateFile.load(path.join(__dirname, './fixtures/state-file.yml'))

			expect(data.value)
				.toBeInstanceOf(Map);
			expect(data.value)
				.toMatchSnapshot();
		});

		it('file does not exist', async () => {
			const result = await StateFile.load('non-existing-file.yml');

			expect(result.isNone())
				.toBeTruthy();
		});

		it('file has invalid structure', async () => {
			const data = await StateFile.load(path.join(__dirname, './fixtures/invalid-state-file.yml'))
				.then(right, left);

			expect(data.value)
				.toMatchSnapshot();
		})
	});

	describe('saving', () => {
		const STATE_FILE_PATH = path.join(__dirname, 'state-file-test.yml');
		afterEach(() => {
			if (fs.existsSync(STATE_FILE_PATH)) {
				fs.unlinkSync(STATE_FILE_PATH);
			}
		});

		it('success', async () => {
			const result = new Map([
				['someproject', 'checksum']
			]);

			await StateFile.save(STATE_FILE_PATH, result);
			const loaded = await StateFile.load(STATE_FILE_PATH);

			expect(loaded)
				.toEqual(just(result));
		})
	});
});
