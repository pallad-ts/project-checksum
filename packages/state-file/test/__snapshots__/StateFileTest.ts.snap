// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateFile loading file has invalid structure 1`] = `[ValidatorError: Invalid state file structure]`;

exports[`StateFile loading success 1`] = `
Map {
  "base" => "somechecksum",
  "another" => "somechecksumforanother",
}
`;

exports[`StateFile validation invalid data 1`] = `
ViolationsList {
  "violations": [
    {
      "message": "must be string",
      "path": [
        "this",
      ],
    },
  ],
}
`;
